core-bundle/src/Cron/Cron.php

```diff
@@ -0,0 +1,176 @@
+    /**
+     * Add a cron service.
+     *
+     * @param object $service
+     */
+    public function addCronJob($service, string $method, string $interval, int $priority = 0, bool $cliOnly = false): void
+    {
+        if (!\in_array($interval, self::INTERVALS, true)) {
+            throw new \InvalidArgumentException(sprintf('Invalid interval "%s"', $interval));
+        }
+
+        $this->crons[$interval][$priority][] = [$service, $method, $cliOnly];
+    }
+
+    /**
+     * Run the registered Contao cron jobs.
+     *
+     * @param bool $cliOnly Whether the cli only crons should be run.
+     */
+    public function run(bool $cliOnly = false): void
+    {
+        // Do not run if the last execution was less than a minute ago
+        if ($this->hasToWait()) {
+            return;
+        }
+
+        $currentRuns = [];
+
+        // Store the current timestamps
+        $currentTimestamps = [
+            'monthly' => date('Ym'),
+            'weekly' => date('YW'),
+            'daily' => date('Ymd'),
+            'hourly' => date('YmdH'),
+            'minutely' => date('YmdHi'),
+        ];
+
+        // Get the timestamps from tl_cron
+        $lastRuns = $this->db->executeQuery("SELECT * FROM tl_cron WHERE name != 'lastrun'")->fetchAll();
+
+        foreach ($lastRuns as $lastRun) {
+            $currentRuns[$lastRun['name']] = $lastRun['value'];
+        }
+
+        // Create the database entries
+        foreach (self::INTERVALS as $interval) {
+            if (!isset($currentRuns[$interval])) {
+                $currentRuns[$interval] = 0;
+                $this->db->insert('tl_cron', ['name' => $interval, 'value' => 0]);
+            }
+        }
+
+        // Run the jobs
+        foreach (self::INTERVALS as $interval) {
+            $currentTimestamp = $currentTimestamps[$interval];
+
+            // Skip empty intervals and jobs that have been executed already
+            if (empty($this->crons[$interval]) || $currentRuns[$interval] === $currentTimestamp) {
+                continue;
+            }
+
+            // Update the database before the jobs are executed, in case one of them fails
+            $this->db->update('tl_cron', ['value' => $currentTimestamp], ['name' => $interval]);
+
+            // Add a log entry if in debug mode (see #4729)
+            if ($this->debug && null !== $this->logger) {
+                $this->logger->log(LogLevel::INFO, 'Running the '.$interval.' cron jobs', ['contao' => new ContaoContext(__METHOD__, TL_CRON)]);
+            }
+
+            // Sort the cron jobs by priority
+            $crons = $this->crons[$interval];
+            krsort($crons);
+            $crons = array_merge(...$crons);
+
+            foreach ($crons as $cron) {
+                // Skip jobs that are only to be run on CLI, when not run via CLI
+                if (!$cliOnly && isset($cron[2]) && true === $cron[2]) {
+                    continue;
+                }
+
+                $cron[0]->{$cron[1]}();
```

Comment:
Can we have a `method_exists()` call here for safety and log on `critical` level, if it does not exist?